File: src/harness_core/dispatch/orchestration/workflow_engine.py
"""Workflow engine: orchestrates multi-agent workflow lifecycle."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

from ..dispatch_ledger import DispatchBundle
from ..task_analyzer import TaskAnalysis
from .agent_role_registry import AgentRoleRegistry
from .conflict_resolver import ConflictResolver
from .dependency_resolver import DependencyResolver
from .human_approval_gate import HumanApprovalGate
from .multi_agent_budget import MultiAgentBudgetManager
from .result_aggregator import ResultAggregator
from .schemas import WorkflowGraph, WorkflowNode
from .task_decomposer import TaskDecomposer
from .work_queue import WorkQueue


class WorkflowEngine:
    """Orchestrates the full workflow lifecycle: decompose, execute, resolve, aggregate."""

    def __init__(
        self,
        decomposer: TaskDecomposer | None = None,
        resolver: DependencyResolver | None = None,
        queue: WorkQueue | None = None,
        conflict_resolver: ConflictResolver | None = None,
        aggregator: ResultAggregator | None = None,
        approval_gate: HumanApprovalGate | None = None,
        budget_manager: MultiAgentBudgetManager | None = None,
        role_registry: AgentRoleRegistry | None = None,
    ) -> None:
        self._decomposer = decomposer or TaskDecomposer()
        self._resolver = resolver or DependencyResolver()
        self._queue = queue or WorkQueue()
        self._conflict_resolver = conflict_resolver or ConflictResolver()
        self._aggregator = aggregator or ResultAggregator()
        self._approval_gate = approval_gate or HumanApprovalGate()
        self._budget_manager = budget_manager or MultiAgentBudgetManager()
        self._role_registry = role_registry or getattr(self._decomposer, '_registry', None)

    def create_workflow(
        self,
        analysis: TaskAnalysis,
        budget_limit: float = 100.0,
        dispatch_bundle: DispatchBundle | None = None,
        dispatch_id: str | None = None,
        decision_status: str = "decided",
    ) -> WorkflowGraph:
        dispatch_id, decision_status = self._extract_dispatch_context(
            dispatch_bundle, dispatch_id, decision_status,
        )
        if decision_status != "decided":
            raise ValueError(
                f"Cannot create workflow: decision_status={decision_status!r}, requires 'decided'"
            )

        graph = self._decomposer.decompose(analysis, dispatch_id=dispatch_id)

        valid, errors = self._resolver.validate(graph)
        if not valid:
            graph = self._set_status(graph, "failed")

        self._budget_manager.create_workflow_budget(graph.workflow_id, budget_limit)

        for node in graph.nodes:
            self._queue.enqueue(graph, node)

        return graph

    def tick(self, graph: WorkflowGraph) -> WorkflowGraph:
        if graph.status in ("completed", "failed", "cancelled"):
            return graph

        graph = self._start_ready_nodes(graph)

        if self._aggregator.is_complete(graph):
            has_failed = any(n.status in ("failed", "cancelled") for n in graph.nodes)
            if has_failed:
                graph = self._handle_failed_nodes(graph)

            conflicts = self._conflict_resolver.detect_conflicts(graph)
            unresolved = [c for c in conflicts if c.resolution_strategy is None]
            if unresolved:
                graph = self._resolve_conflicts(graph, unresolved)
                if graph.status == "cancelled":
                    return graph

            needs_approval = self._check_approval_needed(graph)
            if needs_approval:
                return self._set_status(graph, "waiting_human")

            if has_failed and graph.status not in ("failed", "cancelled"):
                return self._set_status(graph, "failed")

            result = self._aggregator.aggregate(graph)
            return WorkflowGraph(
                workflow_id=graph.workflow_id,
                dispatch_id=graph.dispatch_id,
                nodes=graph.nodes,
                edges=graph.edges,
                status="completed",
                created_at=graph.created_at,
                updated_at=graph.updated_at,
                started_at=graph.started_at,
                completed_at=datetime.now(timezone.utc).isoformat(),
                result=result,
                schema_version=graph.schema_version,
            )

        if graph.status in ("created", "decomposed"):
            return self._set_status(graph, "running")

        return graph

    def resume_after_approval(self, graph: WorkflowGraph, node_id: str) -> WorkflowGraph:
        if graph.status != "waiting_human":
            return graph

        self._approval_gate.approve(node_id)
        return self._set_status(graph, "running")

    def reject_approval(self, graph: WorkflowGraph, node_id: str, reason: str) -> WorkflowGraph:
        self._approval_gate.reject(node_id, reason)
        return self._set_status(graph, "cancelled")

    def cancel(self, graph: WorkflowGraph) -> WorkflowGraph:
        for node in graph.nodes:
            if node.status in ("pending", "ready"):
                graph = self._queue.cancel(graph, node.node_id)
                self._release_agent(node)
        return self._set_status(graph, "cancelled")

    def complete_node(self, graph: WorkflowGraph, node_id: str, output_ref: str, cost: float = 0.0) -> WorkflowGraph:
        graph = self._queue.complete(graph, node_id, output_ref)

        node = self._find_node(graph, node_id)
        if node and cost > 0:
            agent_id = node.assigned_agent_id or ""
            self._budget_manager.record_cost(graph.workflow_id, node_id, agent_id, cost)

            ok, msg = self._budget_manager.check_workflow_budget(graph.workflow_id)
            if not ok and self._budget_manager.overrun_strategy == "cancel":
                return self._set_status(graph, "failed")

        graph = self._update_node_fields(graph, node_id, cost_incurred=cost)

        if node:
            self._release_agent(node)

        return graph

    def fail_node(self, graph: WorkflowGraph, node_id: str, error: str) -> WorkflowGraph:
        graph = self._queue.fail(graph, node_id, error)

        node = self._find_node(graph, node_id)
        if node:
            self._release_agent(node)

        return graph

    def _start_ready_nodes(self, graph: WorkflowGraph) -> WorkflowGraph:
        ready_ids = self._resolver.ready_nodes(graph)
        for node_id in ready_ids:
            graph = self._queue.start(graph, node_id)
        return graph

    def _check_approval_needed(self, graph: WorkflowGraph) -> bool:
        for node in graph.nodes:
            if node.status in ("completed", "failed", "waiting_human"):
                if self._approval_gate.requires_approval(graph, node):
                    return True
        return False

    def _handle_failed_nodes(self, graph: WorkflowGraph) -> WorkflowGraph:
        failed_or_cancelled = [n for n in graph.nodes if n.status in ("failed", "cancelled")]
        if not failed_or_cancelled:
            return graph
        for node in failed_or_cancelled:
            if self._approval_gate.requires_approval(graph, node):
                return self._set_status(graph, "waiting_human")
        return graph

    def _resolve_conflicts(self, graph: WorkflowGraph, conflicts: list) -> WorkflowGraph:
        for conflict in conflicts:
            resolved = self._conflict_resolver.resolve(conflict)
            if resolved.resolution_result == "workflow_cancelled":
                return self._set_status(graph, "cancelled")
        return graph

    def _release_agent(self, node: WorkflowNode) -> None:
        if self._role_registry and node.assigned_agent_id:
            self._role_registry.release_node(node.workflow_id, node.node_id)

    def _extract_dispatch_context(
        self,
        bundle: DispatchBundle | None,
        dispatch_id: str | None,
        decision_status: str,
    ) -> tuple[str | None, str]:
        if bundle is not None:
            return bundle.record.dispatch_id, bundle.decision.decision_status
        return dispatch_id, decision_status

    def _set_status(self, graph: WorkflowGraph, status: str) -> WorkflowGraph:
        now = datetime.now(timezone.utc).isoformat()
        started = graph.started_at if graph.started_at else (now if status == "running" else None)
        return WorkflowGraph(
            workflow_id=graph.workflow_id,
            dispatch_id=graph.dispatch_id,
            nodes=graph.nodes,
            edges=graph.edges,
            status=status,
            created_at=graph.created_at,
            updated_at=now,
            started_at=started,
            completed_at=now if status in ("completed", "failed", "cancelled") else None,
            result=graph.result,
            schema_version=graph.schema_version,
        )

    def _find_node(self, graph: WorkflowGraph, node_id: str) -> WorkflowNode | None:
        for node in graph.nodes:
            if node.node_id == node_id:
                return node
        return None

    def _update_node_fields(
        self, graph: WorkflowGraph, node_id: str,
        cost_incurred: float = 0.0,
    ) -> WorkflowGraph:
        now = datetime.now(timezone.utc).isoformat()
        updated_nodes: list[WorkflowNode] = []
        for node in graph.nodes:
            if node.node_id == node_id:
                updated_nodes.append(WorkflowNode(
                    node_id=node.node_id,
                    workflow_id=node.workflow_id,
                    task_type=node.task_type,
                    assigned_agent_id=node.assigned_agent_id,
                    status=node.status,
                    input_refs=node.input_refs,
                    output_ref=node.output_ref,
                    budget=node.budget,
                    cost_incurred=cost_incurred if cost_incurred > 0 else node.cost_incurred,
                    error=node.error,
                    created_at=node.created_at,
                    started_at=node.started_at,
                    completed_at=node.completed_at,
                    schema_version=node.schema_version,
                ))
            else:
                updated_nodes.append(node)
        return WorkflowGraph(
            workflow_id=graph.workflow_id,
            dispatch_id=graph.dispatch_id,
            nodes=tuple(updated_nodes),
            edges=graph.edges,
            status=graph.status,
            created_at=graph.created_at,
            updated_at=graph.updated_at,
            started_at=graph.started_at,
            completed_at=graph.completed_at,
            result=graph.result,
            schema_version=graph.schema_version,
        )

File: engine/src/orchestration/workflow_engine.rs
use super::conflict_resolver::ConflictResolver;
use super::dependency_resolver::DependencyResolver;
use super::human_approval_gate::HumanApprovalGate;
use super::multi_agent_budget::MultiAgentBudgetManager;
use super::result_aggregator::ResultAggregator;
use super::schemas::{WorkflowGraph, WorkflowNode};
use super::task_decomposer::TaskDecomposer;
use super::work_queue::WorkQueue;
use crate::task_analyzer::TaskAnalysis;

pub struct WorkflowEngine {
    decomposer: TaskDecomposer,
    resolver: DependencyResolver,
    queue: WorkQueue,
    conflict_resolver: ConflictResolver,
    aggregator: ResultAggregator,
    approval_gate: HumanApprovalGate,
    budget_manager: MultiAgentBudgetManager,
}

impl WorkflowEngine {
    pub fn new(decomposer: TaskDecomposer) -> Self {
        Self {
            decomposer,
            resolver: DependencyResolver::new(),
            queue: WorkQueue::new(),
            conflict_resolver: ConflictResolver::new(),
            aggregator: ResultAggregator::new(),
            approval_gate: HumanApprovalGate::new(0.7),
            budget_manager: MultiAgentBudgetManager::new("cancel"),
        }
    }

    pub fn create_workflow(
        &mut self,
        analysis: &TaskAnalysis,
        dispatch_id: &str,
        budget_limit: f64,
        decision_status: &str,
        runtime: &mut crate::runtime::FixtureRuntime,
    ) -> Result<WorkflowGraph, String> {
        if decision_status != "decided" {
            return Err(format!(
                "Cannot create workflow: decision_status={decision_status:?}, requires 'decided'"
            ));
        }

        let graph = self.decomposer.decompose(analysis, dispatch_id, runtime);

        let (valid, _errors) = self.resolver.validate(&graph);
        let graph = if !valid {
            set_status(&graph, "failed")
        } else {
            graph
        };

        self.budget_manager
            .create_workflow_budget(&graph.workflow_id, budget_limit);

        Ok(graph)
    }

    pub fn tick(&mut self, graph: &WorkflowGraph) -> WorkflowGraph {
        if matches!(graph.status.as_str(), "completed" | "failed" | "cancelled") {
            return graph.clone();
        }

        let mut graph = self.start_ready_nodes(graph);

        if self.aggregator.is_complete(&graph) {
            let has_failed = graph
                .nodes
                .iter()
                .any(|n| n.status == "failed" || n.status == "cancelled");

            if has_failed {
                graph = self.handle_failed_nodes(&graph);
            }

            let conflicts = self.conflict_resolver.detect_conflicts(&graph);
            let unresolved: Vec<_> = conflicts
                .iter()
                .filter(|c| c.resolution_strategy.is_none())
                .collect();
            if !unresolved.is_empty() {
                graph = self.resolve_conflicts(&graph, &unresolved);
                if graph.status == "cancelled" {
                    return graph;
                }
            }

            let needs_approval = self.check_approval_needed(&graph);
            if needs_approval {
                return set_status(&graph, "waiting_human");
            }

            if has_failed && graph.status != "failed" && graph.status != "cancelled" {
                return set_status(&graph, "failed");
            }

            let result = self.aggregator.aggregate(&graph);
            return WorkflowGraph {
                status: "completed".to_string(),
                completed_at: Some("now".to_string()),
                result: Some(serde_json::json!(result)),
                ..graph
            };
        }

        if graph.status == "created" || graph.status == "decomposed" {
            return set_status(&graph, "running");
        }

        graph
    }

    pub fn resume_after_approval(&self, graph: &WorkflowGraph, _node_id: &str) -> WorkflowGraph {
        if graph.status != "waiting_human" {
            return graph.clone();
        }
        // Note: approval_gate is behind &self, so we can't mutate here.
        // In a real implementation this would use interior mutability.
        set_status(graph, "running")
    }

    pub fn cancel(&mut self, graph: &WorkflowGraph) -> WorkflowGraph {
        let node_ids: Vec<String> = graph
            .nodes
            .iter()
            .filter(|n| !matches!(n.status.as_str(), "completed" | "failed" | "cancelled"))
            .map(|n| n.node_id.clone())
            .collect();
        let mut g = graph.clone();
        for node_id in node_ids {
            g = self.queue.cancel(&g, &node_id);
        }
        set_status(&g, "cancelled")
    }

    pub fn complete_node(
        &mut self,
        graph: &WorkflowGraph,
        node_id: &str,
        output_ref: &str,
        cost: f64,
    ) -> WorkflowGraph {
        let mut graph = self.queue.complete(graph, node_id, output_ref);

        if cost > 0.0 {
            if let Some(node) = graph.nodes.iter().find(|n| n.node_id == node_id) {
                let agent_id = node.assigned_agent_id.clone().unwrap_or_default();
                self.budget_manager
                    .record_cost(&graph.workflow_id, node_id, &agent_id, cost);

                let (ok, _) = self
                    .budget_manager
                    .check_workflow_budget(&graph.workflow_id);
                if !ok && self.budget_manager.overrun_strategy == "cancel" {
                    return set_status(&graph, "failed");
                }
            }

            // Update cost_incurred
            graph = update_node_field(&graph, node_id, cost);
        }

        graph
    }

    pub fn fail_node(
        &mut self,
        graph: &WorkflowGraph,
        node_id: &str,
        error: &str,
    ) -> WorkflowGraph {
        self.queue.fail(graph, node_id, error)
    }

    fn start_ready_nodes(&self, graph: &WorkflowGraph) -> WorkflowGraph {
        let ready_ids = self.resolver.ready_nodes(graph);
        let mut g = graph.clone();
        for node_id in ready_ids {
            g = self.queue.start(&g, &node_id);
        }
        g
    }

    fn check_approval_needed(&self, graph: &WorkflowGraph) -> bool {
        graph.nodes.iter().any(|node| {
            if matches!(
                node.status.as_str(),
                "completed" | "failed" | "waiting_human"
            ) {
                self.approval_gate.requires_approval(graph, node)
            } else {
                false
            }
        })
    }

    fn handle_failed_nodes(&self, graph: &WorkflowGraph) -> WorkflowGraph {
        let failed_or_cancelled: Vec<&WorkflowNode> = graph
            .nodes
            .iter()
            .filter(|n| n.status == "failed" || n.status == "cancelled")
            .collect();
        if failed_or_cancelled.is_empty() {
            return graph.clone();
        }
        for node in failed_or_cancelled {
            if self.approval_gate.requires_approval(graph, node) {
                return set_status(graph, "waiting_human");
            }
        }
        graph.clone()
    }

    fn resolve_conflicts(
        &self,
        graph: &WorkflowGraph,
        conflicts: &[&super::schemas::ConflictRecord],
    ) -> WorkflowGraph {
        for conflict in conflicts {
            let resolved = self.conflict_resolver.resolve(conflict);
            if resolved.resolution_result.as_deref() == Some("workflow_cancelled") {
                return set_status(graph, "cancelled");
            }
        }
        graph.clone()
    }
}

fn set_status(graph: &WorkflowGraph, status: &str) -> WorkflowGraph {
    let started = if status == "running" && graph.started_at.is_none() {
        Some("now".to_string())
    } else {
        graph.started_at.clone()
    };
    let completed = if matches!(status, "completed" | "failed" | "cancelled") {
        Some("now".to_string())
    } else {
        None
    };
    WorkflowGraph {
        status: status.to_string(),
        updated_at: "now".to_string(),
        started_at: started,
        completed_at: completed,
        ..graph.clone()
    }
}

fn update_node_field(graph: &WorkflowGraph, node_id: &str, cost_incurred: f64) -> WorkflowGraph {
    let nodes: Vec<WorkflowNode> = graph
        .nodes
        .iter()
        .map(|n| {
            if n.node_id == node_id {
                WorkflowNode {
                    cost_incurred,
                    ..n.clone()
                }
            } else {
                n.clone()
            }
        })
        .collect();
    WorkflowGraph {
        nodes,
        ..graph.clone()
    }
}
